[New commondata format] FTDY

scarlehoff opened the pull request:
I'm reopening #1610 here @tgiani
This is just a rebase of your commits there.

These metadata.yaml files however contains some errors (and some things that we might want to change?)

1. The `npoints` field in `hepdata`. Do we need this field @enocera ? (if so I'll add it to the reader)
2. The `theory` field must always come with an `operation` even if it is `NULL`
3. The `FK_tables` are actually a list of list.

So, the theory right now is:

```yaml
theory:
   FK_tables:
     - DYE605
```

but it should be
```yaml
theory:
   FK_tables:
     - - DYE605
   operation: "NULL"
```

(this is due to some FKTables being a concatenation of grids now)

Point 2 is a bit silly in hindsight  but for now it is like that...

scarlehoff commented:
>  I don't know if I have the appetite to reopen this, but I feel the theory metadata has deviated quite a bit from the "we would like experimentalist to be able to write this" goal that we had at some point

What do you mean? I don't think it has changed since we defined it in Milan one year ago?

>  In addition the current definition has a bunch of backwards compatible and "transactional" semantics

like?

Zaharid commented:
> > I don't know if I have the appetite to reopen this, but I feel the theory metadata has deviated quite a bit from the "we would like experimentalist to be able to write this" goal that we had at some point
> 
> What do you mean? I don't think it has changed since we defined it in Milan one year ago?
> 

Probably not, but then again one year has passed.

> > In addition the current definition has a bunch of backwards compatible and "transactional" semantics
> 
> like?

https://github.com/NNPDF/nnpdf/pull/1678/files#diff-eb4ccea229f5af5e39ccb1cc2984780e879f823a50477da80cc65ff859717c8bR112-R115

enocera commented:
> 1. The `npoints` field in `hepdata`. Do we need this field @enocera ? (if so I'll add it to the reader)

This is not required and should be deleted.

scarlehoff commented:
I've updated E605, E866P/R and E906 to follow the new format proposed by @t7phy 

For the name convention I'm following what has been proposed by @cschwan, i.e.,

EXPERIMENT_PROCESS_ENERGY_OBSERVABLE_EXTRA

stopping at the point in which the data is unique. In this case for E605 and E906, the commondata is under the E605/E906 folders. Instead for E866 there were two publications and so the data is under the folders E866_FTDY_800GEV_PXSEC and E866_FTDY_800GEV_PDXSECRATIO. Please @cschwan and @felixhekhorn have a look at the name convention and let me know whether it is correct.

I've also added two keys under `implemented_observables`:

The first is `dataset` which contains the full name of the dataset to be used in runcards. Note that it is at the top of the observable. This doesn't matter for yaml but it might be a nice way of separating the observables.
I think this is useful since otherwise the user needs to "reconstruct" by hand the name which can lead to mistakes (cc @t7phy @enocera I think this would be useful to add in all new commondata).

The second is `process`, which is necessary in order for the name convention to work.

And one top-level key with `energy` (same, it is necessary for the naming convention).

I also think the `file` key under `data_uncertainties` and `data_central` should be removed since it is redundant (these two keys already refer to files).


@tgiani one question, you added also a folder `DYE866P_old_kinematic` but this is in yet a different format. Was this just a leftover? Should I remove it? Or is it a valid variant of DYE866P?

Please, let me know whether all this is correct and whether I can move forward with the reader using these files as target.

cschwan commented:
I'd
- always give at least the experiment and the process (maybe also the energy?). For E605 there are also datasets in which the muon-pair comes from an upsilon resonance, so specifying the process makes it unique again;
- use `DY` instead `FTDY` because the process part is supposed to specify the final state and the experiment/variant usually specify the initial states. Drell-Yan at Tevatron (ppbar) we also call Drell-Yan.

cschwan commented on buildmaster/E605/metadata.yaml:
```diff
@@ -12,17 +12,20 @@ iNSPIRE:
    version: 1
-   tables: [1,2,3,4,5,6,7]
-   npoints: [17,18,18,18,18,18,12] # number of datapoints in each table
+energy: "38.8EV"
```
Should probably be `GEV` instead of `EV`.

t7phy commented:
@scarlehoff if you have a look at https://github.com/NNPDF/nnpdf/blob/more_efficient_metadata_for_new_commondata/buildmaster/CMS_ttBar_8TeV_lj_dif/metadata.yaml#L3 and also have a look at line 23, 48, 73, etc. which contain the observable_name, if you were to concat the line 3 setname with observable_name, you can uniquely identify every observable. In the above example, I will have to capitalize them ofc. In your example, the observable_name would just be empty as the implementation is only with a single observable which is uniquely identified by the setname itself.
 I would therefore suggest to drop the dataset key in every implemented observable, which is redundant (due to setname). I would also suggest dropping of process and energy from every single implemented observable, because again, the setname at the top should explain everything. I do not think it would be a good to over-complicate the reader by requiring to to read so many different keys just to identify a single variable.
Concerning the file key, I guess it helps main consistency with the kinematics part of the metadata block, but I don't have any strong preference there.

scarlehoff commented:
>  I do not think it would be a good to over-complicate the reader by requiring to to read so many different keys just to identify a single variable.

Actually for the reader I would very much like to have all fields: experiment, energy, process and observable. And 4 of these 3 are required anyway (the process is needed for MHOU). Energy is new but I'm guessing it will be necessary in all LHC datasets.

Of course, we can say that the minimum name for the commondata is EXPERIMENT_ENERGY_PROCESS and then it is not necessary to have it in the metadata. That will also be fine. Can we have one paper with two different processes/energies?

t7phy commented:
> > I do not think it would be a good to over-complicate the reader by requiring to to read so many different keys just to identify a single variable.
> 
> Actually for the reader I would very much like to have all fields: experiment, energy, process and observable. And 4 of these 3 are required anyway (the process is needed for MHOU). Energy is new but I'm guessing it will be necessary in all LHC datasets.
> 
> Of course, we can say that the minimum name for the commondata is EXPERIMENT_ENERGY_PROCESS and then it is not necessary to have it in the metadata. That will also be fine. Can we have one paper with two different processes/energies?

Actually I have implemented both cases. Papers with different processes and papers with different energies. In this case we separate the implementations for the processes or energies (suggest by @enocera), so the setname+observable_name can be used for unique identification without any issues.

t7phy commented:
> So I should rename:
> 
> E605 -> E605_DY_38P8GEV E906 -> E906_DY_120GEV
> 
> And if one wants to add another process from E605 then that will be a separate folder even if it happens to be the same paper.
> 
> Did I understand it correctly?

Si

t7phy commented:
I would say even the dataset key in every observable is redundant. All you need is the global key: setname, and the observable specific key: observable_name for unique identification

t7phy commented:
> Ok, I've updated now with all the suggested changes.
> 
> I've left the `dataset` key since in the current naming convention it is necessary due to the following:
> 
> In general the name of the dataset will be: _ and is EXPERIMENT_PROCESS_ENERGY, so the full name is EXPERIMENT_PROCESS_ENERGY_OBSERVABLE
> 
> However, in E866 the folder also includes the observable name, therefore the observable would get added twice. Note that this is also a problem currently in #1684 which is called CMS_TTBAR_8TEV_LJ_DIFF where under this convention `LJ_DIFF` would be the observable however the actual observables still need to concatenate things at the end: `CMS_TTBAR_8TEV_LJ_DIFF_PT`, etc
> 
> What is the proposed solution for this? As I've said before I would like to avoid having keys that appear in some metadata files and not in others regarding the names.
> 
> (I don't have any strong opinion regarding what the correct way should be, I just want to know for sure that I have only one rule to create the name and that I can apply the same rule to all commondata).

As I wrote above, for the E866, the simplest solution is to just have an empty value for observable_name as the setname itself uniquely identifies it. As for the convention, it is EXPERIMENT_PROCESS_ENERGY_MISC.INFO_OBSERVABLENAME, this misc info is important as there are datasets in different papers with slight variations (luminosity for instance). The reader could get the exp name by looking at the what comes before 1st _, process before 2nd _ and so on, however the observable name could be obtained by looking at what comes after the last _. I will need to make lots of changes too, because I usually named observables as M_TTBAR or Y_TTBAR which I will modify to MTTBAR and YTTBAR to ensure this convention.  Does this seem fine?

scarlehoff commented:
Ok, that will basically leave the following convention:

```
<EXPERIMENT>_<PROCESS>_<ENERGY>_<FREE FIELDS>_<OBSERVABLENAME>
```

where `OBSERVABLENAME` is _always_ included in the `metadata.yaml` and if necessary (case of E866) left empty. The free fields can separate only at the topmost level but I think that's ok, basically any possible ambiguity is "casted" into a separation of folders. 

I'm happy with it :D

If @cschwan and @felixhekhorn (also @enocera ofc but I'm assuming he's also ok with it) confirm I would be happy with it (and thus I will update the datasets here and add it to the docs when dealing with #1691).

scarlehoff commented on buildmaster/E605_DY_38P8GEV/metadata.yaml:
```diff
@@ -0,0 +1,50 @@
+   experiment: "DYE605"
+
+implemented_observables:
+  - dataset: "E605_DY_38P8GEV_PXSEC"
```
This will be removed as per the last comment

scarlehoff commented:
@t7phy please have a look at the current version, if everything is ok I'll use these as the "examples" for the reader.

@enocera do you want me to keep `DYE866P_old_kinematic` or can I just remove it?

t7phy commented:
> @t7phy please have a look at the current version, if everything is ok I'll use these as the "examples" for the reader.
> 
> @enocera do you want me to keep `DYE866P_old_kinematic` or can I just remove it?

I think it's good to go.

enocera commented:
> @t7phy please have a look at the current version, if everything is ok I'll use these as the "examples" for the reader.
> 
> @enocera do you want me to keep `DYE866P_old_kinematic` or can I just remove it?

@scarlehoff Let me check what that was about before dropping it, please.

scarlehoff commented:
While one can reconstruct the whole dataset name from the information contained in the foldername + observable...
it is impossible to know what the `foldername` is to be if the `observable` can be included in the name.

I think we need a different rule for cases la E866 where we have:
`E866_DY_800GEV_PXSEC`
and
`E866_DY_800GEV_PDXSECRATIO`

but one could be evil and add as well

`E866_DY_800GEV`

which has inside a `PXSEC` observable. 

Both those cases lead to a dataset called `E866_DY_800GEV_PXSEC`.

I would prefer if we have a convention in which such an ambiguity cannot be generated.

scarlehoff commented:
Also, instead of

```yaml
implemented_observables:
  - observable_name: "ob1"
    ...
  - observable_name: "ob2"
    ...
```

I think it would be better:
```yaml
implemented_observables:
    obs1:
        data: whatever.dat
        plotting: whatever
    obs2:
       ....
```

So that one doesn't need to load all implemented observable in order to look for the desired observable but can rather do `implemented_observables[observable_name]`.

Another question is whether we want to allow for badly constructed observable not to break the good one. So I only load the information for the observable that I actually want so that if `obs2` is broken then as long as I don't ask for `obs2` everything is fine. Or whether instead we want to crash as soon as one observable is not ok.

scarlehoff commented:
The problem with that solution is that it adds a third way of loading the data. I would prefer to have one single way of reading the data. Given a well constructed dataset name I should know exactly where to go to read the data (the old names are not "well constructed" but we cannot do anything about that).

The compatibility dictionary should be limited to the specific case of reading old fits and not be used for anything else. If we want to use a dictionary to tell vp where to load the data from, we can do that, but then all datasets should be read from there (but I don't agree with this).

t7phy commented:
> Also, instead of
> 
> ```yaml
> implemented_observables:
>   - observable_name: "ob1"
>     ...
>   - observable_name: "ob2"
>     ...
> ```
> 
> I think it would be better:
> 
> ```yaml
> implemented_observables:
>     obs1:
>         data: whatever.dat
>         plotting: whatever
>     obs2:
>        ....
> ```
I don't have any opinion on this, if you think this would be better, that's ok with me.
> 
> So that one doesn't need to load all implemented observable in order to look for the desired observable but can rather do `implemented_observables[observable_name]`.
> 
> Another question is whether we want to allow for badly constructed observable not to break the good one. So I only load the information for the observable that I actually want so that if `obs2` is broken then as long as I don't ask for `obs2` everything is fine. Or whether instead we want to crash as soon as one observable is not ok.

I guess better to load what is needed only. The whole idea of this new metadata format is to be able to implement other observables whenever we want, however it would be bad if a new implementation breaks something already implemented.

t7phy commented:
> The problem with that solution is that it adds a third way of loading the data. I would prefer to have one single way of reading the data. Given a well constructed dataset name I should know exactly where to go to read the data (the old names are not "well constructed" but we cannot do anything about that).
> 
> The compatibility dictionary should be limited to the specific case of reading old fits and not be used for anything else. If we want to use a dictionary to tell vp where to load the data from, we can do that, but then all datasets should be read from there (but I don't agree with this).

Although I agree that in principal we should have a consistent way we read them, as we know, this particular set is an exception so I guess we could make an exception for this case (and if so any comes in the future where we use the dictionary, which is actually meant for something else, to help here). I am not sure of the problem here.

scarlehoff commented:
While this problem is exposed by this specific dataset, the ambiguity is completely general. I would prefer one consistent way that works for all datasets we already have.

> The whole idea of this new metadata format is to be able to implement other observables whenever we want, however it would be bad if a new implementation breaks something already implemented.

I guess the advantage of being strict is that we ensure that we are not using commondata with broken observables (which is probably a good thing)

scarlehoff commented:
(on the other hand, by being strict the whole commondata is loaded and then it doesn't matter whether we choose the `list` or the `dictionary` approach... so I guess I'll go for the strict checking so that the datasets which are already implemented are ok)

t7phy commented:
> While this problem is exposed by this specific dataset, the ambiguity is completely general. I would prefer one consistent way that works for all datasets we already have.
> 
Well then the only other option I see is the option I had mentioned the other day in case of this dataset. We all the references to be a list, i.e. multiple arxiv, inspire and hepdata sources. This way if a dataset is provided to us from multiple sources even though it is for the same experiment/process/etc., it can still be in the same folder. Then PXSEC and PDXSECRATIO become observable names as we would like them to be. The ambiguity goes away because then we have a strict convention that there must be observable name, and it can't be left empty.

> > The whole idea of this new metadata format is to be able to implement other observables whenever we want, however it would be bad if a new implementation breaks something already implemented.
> 
> I guess the advantage of being strict is that we ensure that we are not using commondata with broken observables (which is probably a good thing)
Well technically, the new implementations would go through PR reviews and not really be broken unless someone merges without review :P

scarlehoff commented:
But at that point, wouldn't it be easier to be strict with the convention, i.e., `<folder_name>_<observable>` and in this case and other which might arise add `_extras`. Would that be fine?

t7phy commented:
> But at that point, wouldn't it be easier to be strict with the convention, i.e., `<folder_name>_<observable>` and in this case and other which might arise add `_extras`. Would that be fine?

I am not sure I understand what you mean here

t7phy commented:
> But at that point, wouldn't it be easier to be strict with the convention, i.e., `<folder_name>_<observable>` and in this case and other which might arise add `_extras`. Would that be fine?

Yes, `<Exp.>_<Process>_<Energy>_<Extras>_<observable>` whereby  `<Exp.>_<Process>_<Energy>_<Extras>` is the folder name.

scarlehoff commented:
I've update the names of the two E866 observables. I've also disconnected this branch from the reader (it's easy enough to copy the relevant folders every time I want to test the reader).

scarlehoff commented:
@enocera @tgiani I think the `MULT` and `ADD` systematics uncertainties are swapped (at least in the E605 dataset). Do to that I think the ADD uncertainties are computed wrong (since I guess the single ADD is a combination of all the other ADDs).

However, I don't see the 'add in quadrature' operation in the `filter.py` (and there are not many comments there...) so I cannot help but wonder... why are the number of systematics uncertainties different in the old and new metadata? Which one of the two is bugged?

enocera commented:
> @enocera @tgiani I think the `MULT` and `ADD` systematics uncertainties are swapped (at least in the E605 dataset). Do to that I think the ADD uncertainties are computed wrong (since I guess the single ADD is a combination of all the other ADDs).
> 
> However, I don't see the 'add in quadrature' operation in the `filter.py` (and there are not many comments there...) so I cannot help but wonder... why are the number of systematics uncertainties different in the old and new metadata? Which one of the two is bugged?

@scarlehoff Let me also check this. However, as I mentioned in #1684 may it be that the difference in the number of sys uncertainties is that we only have absolute uncs (in the new format) instead of absolute and relative (in the old format)?

enocera commented:
> > @t7phy please have a look at the current version, if everything is ok I'll use these as the "examples" for the reader.
> > @enocera do you want me to keep `DYE866P_old_kinematic` or can I just remove it?
> 
> @scarlehoff Let me check what that was about before dropping it, please.

You can drop it.

enocera commented:
> Ah, ok! I thought they should match the old ones. Are these kinds of differences always expected? To know what level of accuracy should I expect when testing... I can just do "ok, if it's ok at 5% then I'm ok with it". I would like to automatize this testing as much as possible.

For old data sets (that is data sets NOT taken from Hepdata) these differences can happen.

scarlehoff commented:
@enocera follow up questions

- DYE866P is completely different. I guess this is the reason an older ersion was also done. Do you know the reason for these differences? So that I can add them as comments.

<img width="805" alt="Screenshot 2023-03-31 at 11 07 56" src="https://user-images.githubusercontent.com/20564015/229077455-bbf0ed1d-f46d-4740-b3ee-c1959695cdb3.png">

For now the differences for this branch that I don't understand are:

- DYE605: a factor of two in the additive uncertainties.
- DYE886P: see above
- DYE886R: only kinematics are different
- DYE906R: the number of uncertainties included is very different

Kinematics are always different. I'll redo the `kinematics.yaml` table so that it is the same as the old ones.

Of course, this is the "raw" datasets... but all the `dw` (or `dw_ite`) are still to be done, right?

enocera commented:
> @enocera follow up questions
> 
>     * DYE866P is completely different. I guess this is the reason an older ersion was also done. Do you know the reason for these differences? So that I can add them as comments.

Not on top of my head. Let me check.

> For now the differences for this branch that I don't understand are:
> 
>     * DYE605: a factor of two in the additive uncertainties.
> 
>     * DYE886P: see above
> 
>     * DYE886R: only kinematics are different
> 
>     * DYE906R: there is a factor of 1 for some of the uncertainties
> 

Let me check all these.

> Of course, this is the "raw" datasets... but all the `dw` (or `dw_ite`) are still to be done, right?

Correct. And that's because I told @tgiani to ignore variants for the time being. Let me implement these variants.

scarlehoff commented:
Just to clarify, the factor of 1 is actually of -1 and it happens for `sys_4` and `sys_5` and I think they are to the ambiguity of taking + or - when taking the square root (so at the end of the day the produced covmat would be the same).
The question is whether this differences between old and new are ok.

scarlehoff commented:
Hi @enocera, are these datasets finished? (I remember you mentioned there were some issues that you might want to update, but I don't remember exactly what was the situation...)

If these are finished I was thinking of updating them with the plotting options etc to start collecting datasets in https://github.com/NNPDF/nnpdf/pull/1813

scarlehoff commented:
At this point this is very much obsolete. The information is not lost in case it is necessary so I'm just closing the PR.
